### PR TITLE
WIP smup

### PR DESCRIPTION
**Pull request by fitzgen**
So this compiles and stuff, but not yet passing tests.

Here are the results of the DOM wpt, which I haven't started digging into yet:

```
Ran 285 tests finished in 104.0 seconds.
  • 21 ran as expected. 15 tests skipped.
  • 31 tests crashed unexpectedly
  • 233 tests timed out unexpectedly
```

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17143)

### CODE REVIEW

**Comment by fitzgen** [comment]
Fixed some issues in CodegenRust.py causing bad `*const jsapi::JSJitInfo` pointers; now seeing some `MOZ_CRASH`es inside spidermonkey.

**Comment by fitzgen** [comment]
Ok, fixed some more things to do with proxies, which upstream SpiderMonkey had changed where expandos and private objects are stored and I got subtly wrong when rebasing.

New dom WPT results are pretty close to passing:

```
Ran 285 tests finished in 81.0 seconds.
  • 248 ran as expected. 15 tests skipped.
  • 7 tests crashed unexpectedly
  • 30 tests timed out unexpectedly
```

**Comment by fitzgen** [comment]
No more crashes!

```
Ran 285 tests finished in 79.0 seconds.
  • 272 ran as expected. 15 tests skipped.
  • 13 tests timed out unexpectedly
  • 7 tests had unexpected subtest results
```

**Comment by fitzgen** [comment]
Seems like the timeouts and unexpected results were because I was using a `debug + debugmozjs` build \o/

```
Ran 285 tests finished in 26.0 seconds.
  • 285 ran as expected. 15 tests skipped.
```

**Comment by KiChjang** [comment]
Huh, autospider is requiring gcc 4.9 or newer, and travis CI uses 4.8.4...

**Comment by jdm** [comment]
https://gist.github.com/cotsog/3ce84675af0d74438d91 may help.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -2948,20 +2944,20 @@ def defineAliasesFor(m):
...
-    JS_NewObjectWithoutMetadata(cx, %(holderClass)s, %(holderProto)s));
+    jsapi::JS_NewObjectWithoutMetadata(cx, %(holderClass)s, %(holderProto)s));
 assert!(!unforgeable_holder.is_null());
 """ % {'holderClass': holderClass, 'holderProto': holderProto}))
             code.append(InitUnforgeablePropertiesOnHolder(self.descriptor, self.properties))
             code.append(CGGeneric("""\
-JS_SetReservedSlot(prototype.get(), DOM_PROTO_UNFORGEABLE_HOLDER_SLOT,
-                   ObjectValue(unforgeable_holder.get()))"""))
+jsapi::JS_SetReservedSlot(prototype.get(), DOM_PROTO_UNFORGEABLE_HOLDER_SLOT,
+                   &ObjectValue(unforgeable_holder.get()) as *const _)"""))
```
Comment (jdm) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Unnecessary cast.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 2]
```diff
@@ -3642,23 +3638,33 @@ def defineJitInfo(self, infoName, opName, opType, infallible, movable,
...
-                            ${slotIndex} as u16,
-                        )
+                jsapi::JSJitInfo {
+                    __bindgen_anon_1: jsapi::JSJitInfo__bindgen_ty_1 {
+                        method: unsafe {
+                            mem::transmute(${opName} as
+                                unsafe extern "C" fn(*mut jsapi::JSContext,
+                                                     jsapi::JS::HandleObject,
+                                                     _,
+                                                     _) -> bool) },
```
Comment (jdm) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Why is this transmute required?

File: components/script/dom/bindings/conversions.rs [hunk 3]
```diff
@@ -23,7 +23,7 @@
 //! | DOMString               | `DOMString`                      |
 //! | USVString               | `USVString`                      |
 //! | ByteString              | `ByteString`                     |
-//! | object                  | `*mut JSObject`                  |
+//! | object                  | `*mut jsapi::JSObject`                  |
```
Comment (jdm) [inline components/script/dom/bindings/conversions.rs]: nit: the alignment in this table is now off.

File: components/script/dom/bindings/error.rs [hunk 4]
```diff
@@ -146,43 +139,37 @@ pub struct ErrorInfo {
...
+        let lineno = 0;
+        let column = 0;
 
         let message = {
-            let message = (*report).ucmessage;
+            let message = (*report)._base.message_.data_;
             let length = (0..).find(|idx| *message.offset(*idx) == 0).unwrap();
-            let message = from_raw_parts(message, length as usize);
-            String::from_utf16_lossy(message)
+            let message = from_raw_parts(message as _, length as usize);
```
Comment (jdm) [inline components/script/dom/bindings/error.rs]: This cast looks surprising.

File: components/script/dom/bindings/interface.rs [hunk 5]
```diff
@@ -197,40 +186,43 @@ impl InterfaceConstructorBehavior {
...
-                                OnNewGlobalHookOption::DontFireOnNewGlobalHook,
+                                jsapi::JS::OnNewGlobalHookOption::DontFireOnNewGlobalHook,
                                 &options));
     assert!(!rval.is_null());
 
     // Initialize the reserved slots before doing anything that can GC, to
     // avoid getting trace hooks called on a partially initialized object.
-    JS_SetReservedSlot(rval.get(), DOM_OBJECT_SLOT, PrivateValue(private));
+    let private = PrivateValue(private);
+    jsapi::JS_SetReservedSlot(rval.get(), DOM_OBJECT_SLOT, &private as *const _);
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: Unnecessary cast.

File: components/script/dom/bindings/interface.rs [hunk 6]
```diff
@@ -197,40 +186,43 @@ impl InterfaceConstructorBehavior {
...
-                       DOM_PROTOTYPE_SLOT,
-                       PrivateValue(Box::into_raw(proto_array) as *const libc::c_void));
+        box [0 as *mut jsapi::JSObject; PrototypeList::PROTO_OR_IFACE_LENGTH];
 
-    let _ac = JSAutoCompartment::new(cx, rval.get());
-    JS_FireOnNewGlobalObject(cx, rval.handle());
+    let proto_array = PrivateValue(Box::into_raw(proto_array) as *const libc::c_void);
+    jsapi::JS_SetReservedSlot(rval.get(),
+                              DOM_PROTOTYPE_SLOT,
+                              &proto_array as *const _);
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: Unnecessary cast.

File: components/script/dom/bindings/weakref.rs [hunk 7]
```diff
@@ -60,7 +60,8 @@ pub trait WeakReferenceable: DomObject + Sized {
                     count: Cell::new(1),
                     value: Cell::new(Some(NonZero::new(self))),
                 });
-                JS_SetReservedSlot(object, DOM_WEAK_SLOT, PrivateValue(ptr as *const c_void));
+                let ptr = PrivateValue(ptr as *const c_void);
+                jsapi::JS_SetReservedSlot(object, DOM_WEAK_SLOT, &ptr as *const _);
```
Comment (jdm) [inline components/script/dom/bindings/weakref.rs]: Unnecessary cast.

File: components/script/dom/windowproxy.rs [hunk 8]
```diff
@@ -123,10 +114,11 @@ impl WindowProxy {
 
             // The window proxy owns the browsing context.
             // When we finalize the window proxy, it drops the browsing context it owns.
-            SetProxyExtra(js_proxy.get(), 0, &PrivateValue((&*window_proxy).as_void_ptr()));
+            let private = js::jsval::PrivateValue((&*window_proxy).as_void_ptr());
+            SetProxyReservedSlot(js_proxy.get(), 0, &private as *const _);
```
Comment (jdm) [inline components/script/dom/windowproxy.rs]: Unnecessary cast.

File: components/script/dom/windowproxy.rs [hunk 9]
```diff
@@ -159,19 +151,19 @@ impl WindowProxy {
             let window = DissimilarOriginWindow::new(global_to_clone_from, &*window_proxy);
             let window_jsobject = window.reflector().get_jsobject();
             assert!(!window_jsobject.get().is_null());
-            assert!(((*get_object_class(window_jsobject.get())).flags & JSCLASS_IS_GLOBAL) != 0);
-            let _ac = JSAutoCompartment::new(cx, window_jsobject.get());
+            assert!(((*js::rust::get_object_class(window_jsobject.get())).flags & jsapi::JSCLASS_IS_GLOBAL) != 0);
```
Comment (jdm) [inline components/script/dom/windowproxy.rs]: Perhaps we should use the new helper that was added for this.

File: components/script/script_runtime.rs [hunk 10]
```diff
@@ -233,89 +229,84 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
             false
         };
         let mode = if val {
-            JSGCMode::JSGC_MODE_INCREMENTAL
+            jsapi::JSGCMode::JSGC_MODE_INCREMENTAL
         } else if compartment {
```
Comment (jdm) [inline components/script/script_runtime.rs]: Perhaps rename this now.

**Review by jdm — CHANGES_REQUESTED** [review]
All pretty minor changes that I've suggested.

File: components/script/dom/bindings/error.rs (hunk 4 above)
Comment (fitzgen) [reply components/script/dom/bindings/error.rs]: Without the cast we get

```
error[E0308]: mismatched types
   --> /home/fitzgen/servo/components/script/dom/bindings/error.rs:161:37
    |
161 |             String::from_utf8_lossy(message)
    |                                     ^^^^^^^ expected u8, found i8
    |
    = note: expected type `&[u8]`
               found type `&[i8]`
```

File: components/script/dom/bindings/codegen/CodegenRust.py (hunk 2 above)
Comment (fitzgen) [reply components/script/dom/bindings/codegen/CodegenRust.py]: It's there to change the `*const ActualType` parameter into a `*const std::os::raw::c_void`. Upon reflection this is better done with `as`.

Comment (fitzgen) [reply components/script/dom/bindings/codegen/CodegenRust.py]: hmmm nevermind, `as` doesn't work here:

```
error[E0605]: non-primitive cast: `unsafe extern "C" fn(*mut js::jsapi::JSContext, js::jsapi::JS::Handle<*mut js::jsapi::JSObject>, *const dom::xmlhttprequesteventtarget::XMLHttpRequestEventTarget, js::jsapi::JSJitSetterCallArgs) -> bool {dom::bindings::codegen::Bindings::XMLHttpRequestEventTargetBinding::XMLHttpRequestEventTargetBinding::set_onloadend}` as `unsafe extern "C" fn(*mut js::jsapi::JSContext, js::jsapi::JS::Handle<*mut js::jsapi::JSObject>, *mut std::os::raw::c_void, *const js::jsapi::JSJitMethodCallArgs) -> bool`
   --> /home/fitzgen/servo/target/debug/build/script-11a815433a47bb39/out/Bindings/XMLHttpRequestEventTargetBinding.rs:885:22
    |
885 |           method: Some(set_onloadend as
    |  ______________________^
886 | |             unsafe extern "C" fn(*mut jsapi::JSContext,
887 | |                                  jsapi::JS::Handle<*mut jsapi::JSObject>,
888 | |                                  _,
889 | |                                  _) -> bool),
    | |___________________________________________^
    |
    = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait
```

**Comment by fitzgen** [comment]
Feedback addressed. Need to rebase again. Still blocked on windows stuff in https://github.com/servo/mozjs/pull/119.

**Review by jdm — APPROVED** [review]

**Comment by nox** [comment]
Could you please revert all the qualified paths to `jsapi` items you introduced?

**Comment by fabricedesre** [comment]
Is anyone still working on that?

**Comment by jdm** [comment]
@fabricedesre @asajeffrey is going to be resuming this work.

**Comment by asajeffrey** [comment]
I'm going to talk to @nox in Berlin about how to move forward; I'm happy to look at smup, but the problem is that I have approximately zero windows experience.

**Comment by nox** [comment]
@asajeffrey For Windows support I haven't much to say but "good luck".

**Comment by asajeffrey** [comment]
Er, thanks.

**Comment by jdm** [comment]
Superseded.

**Pull request closed without merging**